lightning/src/events/mod.rs

```diff
@@ -203,11 +204,18 @@ impl PaymentPurpose {
 					payment_context: context,
 				})
 			},
-			Some(PaymentContext::AsyncBolt12Offer(_context)) => {
-				// This code will change to return Self::Bolt12OfferPayment when we add support for async
-				// receive.
-				Err(())
-			},
+			Some(PaymentContext::AsyncBolt12Offer(_)) => {
+				let invoice_request = invreq.ok_or(())?;
+				if payment_preimage.is_none() { return Err(()) }
+				Ok(PaymentPurpose::Bolt12OfferPayment {
+					payment_preimage,
+					payment_secret,
+					payment_context: Bolt12OfferContext {
+						offer_id: invoice_request.offer_id,
+						invoice_request: invoice_request.fields(),
+					},
+				})
```

Comment:
Wonder if we should pass in the appropriate `Bolt12OfferContext` instead of adding the `invreq` parameter to `from_parts`.

Reply:
Hmm, I don't really have a strong feeling there. It seems nice to keep more code out of `channelmanager.rs` by default. Do you have a preference or is there some tradeoff I'm missing?

Reply:
You wouldn't need to pass in `None` everywhere else.

Reply:
I tried out this approach but it wasn't obviously an improvement IMO. Another option is to add a dedicated `PaymentPurpose` for async payments. I think we decided against that in the past but could revisit when the feature is more fully supported/exposed in the public API as well.

Reply:
Hmm... it doesn't feel right that we pass it in and ignore it all cases but one. All the other parameters are used or checked except for that one. So you could accidentally pass it in with the wrong context and it will be ignored.